File: backend_python/src/middleware/proxy_auth.py
import json
from datetime import datetime
from typing import Optional

from fastapi import HTTPException, Request
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer

from src.db.database import db


class TokenData:
    """Token data from database"""

    def __init__(self, token_id: int, server_id: int, permissions: Optional[list[str]] = None):
        self.token_id = token_id
        self.server_id = server_id
        self.permissions = permissions or []


security = HTTPBearer()


async def verify_token(credentials: HTTPAuthorizationCredentials, server_id: int) -> TokenData:
    """Verify access token and check permissions"""
    token = credentials.credentials

    # Query token from database
    row = db.fetchone(
        """
        SELECT id, server_id, permissions, expires_at
        FROM mcp_access_tokens
        WHERE token = ? AND server_id = ?
        """,
        (token, server_id),
    )

    if not row:
        raise HTTPException(status_code=401, detail="Invalid token")

    # Check expiration
    if row["expires_at"]:
        expires_at = datetime.fromisoformat(row["expires_at"])
        if datetime.now() > expires_at:
            raise HTTPException(status_code=401, detail="Token expired")

    # Parse permissions
    permissions = []
    if row["permissions"]:
        try:
            permissions = json.loads(row["permissions"])
        except Exception:
            pass

    return TokenData(token_id=row["id"], server_id=row["server_id"], permissions=permissions)


def check_permission(token_data: TokenData, required_permission: str):
    """Check if token has required permission"""
    if required_permission not in token_data.permissions:
        raise HTTPException(status_code=403, detail=f"Permission '{required_permission}' required")

File: backend/src/middleware/proxyAuth.ts
import { Request, Response, NextFunction } from 'express';
import Database from 'better-sqlite3';
import { join } from 'path';

const dbPath = join(__dirname, '../../data/mcp-market.db');
const db = new Database(dbPath);

export interface AuthenticatedRequest extends Request {
  tokenInfo?: {
    id: number;
    token: string;
    server_id: number;
    name: string | null;
    permissions: string[];
    expires_at: string | null;
  };
}

export const proxyAuth = (req: AuthenticatedRequest, res: Response, next: NextFunction) => {
  try {
    // Extract token from Authorization header
    const authHeader = req.headers.authorization;
    if (!authHeader || !authHeader.startsWith('Bearer ')) {
      return res.status(401).json({ error: 'Missing or invalid Authorization header' });
    }

    const token = authHeader.substring(7); // Remove 'Bearer ' prefix

    // Query token from database
    const stmt = db.prepare(`
      SELECT id, token, server_id, name, permissions, expires_at
      FROM mcp_access_tokens
      WHERE token = ?
    `);
    const tokenInfo = stmt.get(token) as any;

    if (!tokenInfo) {
      return res.status(401).json({ error: 'Invalid token' });
    }

    // Check if token is expired
    if (tokenInfo.expires_at) {
      const expiresAt = new Date(tokenInfo.expires_at);
      if (expiresAt < new Date()) {
        return res.status(401).json({ error: 'Token expired' });
      }
    }

    // Parse permissions
    const permissions = tokenInfo.permissions ? JSON.parse(tokenInfo.permissions) : [];

    // Attach token info to request
    req.tokenInfo = {
      ...tokenInfo,
      permissions,
    };

    next();
  } catch (error) {
    console.error('Auth middleware error:', error);
    res.status(500).json({ error: 'Authentication failed' });
  }
};

// Check if token has specific permission
export const requirePermission = (permission: string) => {
  return (req: AuthenticatedRequest, res: Response, next: NextFunction) => {
    if (!req.tokenInfo) {
      return res.status(401).json({ error: 'Not authenticated' });
    }

    if (!req.tokenInfo.permissions.includes(permission)) {
      return res.status(403).json({ error: `Permission '${permission}' required` });
    }

    next();
  };
};
